src/image/detail/MaskDict.cc

```diff
@@ -0,0 +1,181 @@
+    std::shared_ptr<MaskDict> dict = GlobalState::get().defaultMaskDict;
+    if (!mpd.empty()) {
+        dict = std::shared_ptr<MaskDict>(new MaskDict(mpd));
+        GlobalState::get().allMaskDicts.insert(dict.get());
+    }
+    return dict;
+}
+
+std::shared_ptr<MaskDict> MaskDict::setDefaultDict(std::shared_ptr<MaskDict> dict) {
+    return GlobalState::get().defaultMaskDict = std::move(dict);
+}
+
+std::shared_ptr<MaskDict> MaskDict::detachDefault() {
+    GlobalState::get().defaultMaskDict = GlobalState::get().defaultMaskDict->clone();
+    return GlobalState::get().defaultMaskDict;
+}
+
+void MaskDict::addMaskPlane(std::string const & name, int bitId) {
+    for (MaskDict * dict : GlobalState::get().allMaskDicts) {
+        // is bitId already in use?
+        auto const it = std::find_if(
+            dict->begin(), dict->end(),
+            [bitId](auto const & item) { return item.second == bitId; }
+        );
+        if (it != dict->end()) {
+            continue;
+        }
+        // is name already in use?
+        if (dict->find(name) == dict->end()) {
+            dict->add(name, bitId);
+        }
+    }
+}
+
+MaskDict::~MaskDict() { GlobalState::get().allMaskDicts.erase(this); }
+
+std::shared_ptr<MaskDict> MaskDict::clone() const {
+    std::shared_ptr<MaskDict> dict(new MaskDict(*this));
+    GlobalState::get().allMaskDicts.insert(dict.get());
+    return dict;
+}
+
+int MaskDict::getUnusedPlane() const {
+    if (empty()) {
+        return 0;
+    }
+
+    auto const it = std::max_element(
+        begin(), end(),
+        [](auto const & a, auto const & b) { return a.second < b.second; }
+    );
+    assert(it != end());
+    int id = it->second + 1;  // The maskPlane to use if there are no gaps
+
+    for (int i = 0; i < id; ++i) {
+        // is i already used in this Mask?
+        auto const it = std::find_if(
+            begin(), end(),
+            [i](auto const & item) { return item.second == i; }
+        );
+        if (it == end()) {  // Not used; so we'll use it
+            return i;
+        }
+    }
+
+    return id;
+}
+
+int MaskDict::getMaskPlane(std::string const & name) const {
+    auto i = find(name);
+    return (i == end()) ? -1 : i->second;
+}
+
+void MaskDict::print() const {
+    for (auto const & item : *this) {
+        std::cout << "Plane " << item.second << " -> " << item.first << std::endl;
+    }
+}
+
+void MaskDict::add(std::string const & name, int bitId) {
+    _dict[name] = bitId;
+    _hash = _calcHash();
+}
+
+void MaskDict::erase(std::string const & name) {
+    if (_dict.find(name) != _dict.end()) {
+        _dict.erase(name);
+        _hash = _calcHash();
+    }
+}
+
+void MaskDict::clear() {
+    _dict.clear();
+    _hash = 0x0;
+}
+
+std::size_t MaskDict::_calcHash() const {
+    std::size_t hash = 0x0;
+    for (auto const & item : _dict) {
+        hash = (hash << 1) ^
+                boost::hash<std::string>()(item.first + str(boost::format("%d") % item.second));
+    }
```

Comment:
Huh.  I have no idea; just inherited this logic from what was there before.  I'll be brave on this one and try to change it (and maybe use std::hash, too).

Reply:
Unfortunately, the standard collections [don't have standard `hash` specializations](https://en.cppreference.com/w/cpp/utility/hash#Standard_specializations_for_library_types). It's a rather strange oversight, precisely because of classes like `Mask` and `MaskDict`.